src/compiler/program.ts

```diff
@@ -36,57 +36,369 @@ namespace ts {
+    // that contains import. However sometimes it might  be useful to apply mappings to relative module names as well.
+    // Since the there might be different relative file names that refer to the same file we need to convert relative module name to non-relative 
+    // so we can just apply path mapping to it. This conversion can be performed using 'rootDirs' compiler option - this options stores list of root directories
+    // for the project (if root directory is not absolute it will be converted to absolute using baseUrl as a base path).
+    // First - we convert relative module name to absolute using location of file that contains import. 
+    // Then we try to find what element is rootDirs is the longest prefix for this absolute file name (respecting case):
+    //      <absoluteFileName> = <longestRootDir><suffix> 
+    // Once it is done - value of <suffix> is a be non-relative name for initially relative module name.
+    // Note that if element in rootDirs should always end with '/' so it will be appended automatically if value in configuration does not have it. 
+    // For example:
+    // baseUrl: /a/b/c
+    // "paths": {
+    //     "*": [ "*", "generated/*" ] // try to resolve module name as '/a/b/c/<name>' then look it in '/a/b/c/generated/<name>'
+    // }
+    // "rootDirs": [
+    //     "./src",      // root dir is project baseUrl/src
+    //     "./generated" // root dir is baseUrl/generated
+    //     "/e/f"        // root dir is /e/f 
+    // ]
+    // @file: /a/b/c/src/form1.ts
+    //     import {x} from "./form.content.ts"
+    // @file: /a/b/c/generated/form1.content.ts
+    //     export var x = ...
+    // first './form.content.ts' needs to be converted to non-relative name.
+    // 1. convert it to absolute name '/a/b/c/src/form1.content.ts' ->
+    //    find longest prefix in rootDirs that match this path, it will be './src' (after expansion '/a/b/c/src/') ->
+    //    suffix is form1.content.ts
+    // 2. apply path mappings to 'form1.content.ts' ->
+    //    '*' pattern will be matched ->
+    //    '*' substitution yields '/a/b/c/form1.content.ts' - file does not exists
+    //    'generated/*' substitution yields '/a/b/c/generated/form1.content' - file exists - OK
+    export function baseUrlModuleNameResolver(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModuleWithFailedLookupLocations {
+        const baseUrl = compilerOptions.baseUrl !== undefined ? compilerOptions.baseUrl : compilerOptions.inferredBaseUrl;
+        Debug.assert(baseUrl !== undefined);
+
+        const supportedExtensions = getSupportedExtensions(compilerOptions);
+        const traceEnabled = isTraceEnabled(compilerOptions, host);
+
+        if (isRootedDiskPath(moduleName)) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.Resolving_rooted_module_name_0_use_it_as_a_candidate_location, moduleName);

```

Comment:
Should the `traceEnabled` checked be in the trace function since there are decent numbers of them around?